### PR TITLE
[WIP][RFC] Concept for connection timeout

### PR DESCRIPTION
**Pull request by cboden**
The implementation for a connection timeout. Any connector could use the method internally to create a connection timeout.

### CODE REVIEW

**Comment by WyriHaximus** [comment]
@cboden are you planning on updating this PR soon or want me to make a PR to your PR using `CancellablePromiseInterface`?

**Comment by cboden** [comment]
@WyriHaximus Help is always welcome  :-)

**Comment by WyriHaximus** [comment]
@cboden alright will look into it :).

**Comment by cboden** [comment]
Added cancellable. Still WIP, API needs adjusting, needs tests. I'm not sure if the trait belongs in this library, maybe Stream instead as it isn't solely tied to a connecting client.

**Comment by clue** [comment]
:+1: for pushing this forward, though I'm not sure I like the API and how it's intertwined with the "default" `Connector`.

IMO we should first figure out some use cases for this feature to see how this is actually going to be used.

The socket-client component is quite lowlevel, so it's unlikely to be used by many users directly. It's probably more likely going to be used as part of some high level abstractions (say, a Redis or MySQL client implementation). These high level components depend on a `ConnectorInterface` and will likely use DI to inject this. They probably care little about *how* the connector works. Neither should they care about any "additional" parameters. They *do* care about the target (hostname and port). Hence (IMHO) they should not have to take care of connection timeouts.

>  I'm not sure if the trait belongs in this library

I agree that this should probably be moved to a new component instead (promise-timeout or just timeout perhaps?).

File: src/TimeoutTrait.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace React\SocketClient;
+
+use React\EventLoop\LoopInterface;
+use React\Promise\CancellablePromiseInterface;
+
+trait TimeoutTrait
+{
+    protected function setTimeout(LoopInterface $loop, CancellablePromiseInterface $promise, $seconds = 30)
```
Comment (Renegade334) [inline src/TimeoutTrait.php]: Is much gained from having a default value for `$seconds` here? Any value is pretty arbitrary.

**Pull request closed without merging**